crates/qasm3/src/exporter.rs

```diff
@@ -0,0 +1,1412 @@
+            RegisterType::ClassicalRegister(classical_register) => classical_register
+                .bits()
+                .map(BitType::ShareableClbit)
+                .collect(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Counter {
+    current: usize,
+}
+
+impl Counter {
+    fn new() -> Self {
+        Self { current: 0 }
+    }
+}
+
+impl Iterator for Counter {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let n = self.current;
+        self.current = self.current.wrapping_add(1);
+        Some(n)
+    }
+}
+
+struct BuildScope {
+    circuit_data: CircuitData,
+    bit_map: HashMap<BitType, BitType>,
+}
+
+impl BuildScope {
+    fn new(
+        circuit_data: &CircuitData,
+        qubits: &[ShareableQubit],
+        clbits: &[ShareableClbit],
+    ) -> Self {
+        let mut bit_map: HashMap<BitType, BitType> = HashMap::new();
+        for q in qubits.iter() {
+            bit_map.insert(
+                BitType::ShareableQubit(q.clone()),
+                BitType::ShareableQubit(q.clone()),
+            );
+        }
+
+        for c in clbits.iter() {
+            bit_map.insert(
+                BitType::ShareableClbit(c.clone()),
+                BitType::ShareableClbit(c.clone()),
+            );
+        }
+        Self {
+            circuit_data: circuit_data.clone(),
+            bit_map,
+        }
+    }
+
+    fn with_mappings(circuit_data: CircuitData, bit_map: HashMap<BitType, BitType>) -> Self {
+        Self {
+            circuit_data,
+            bit_map,
+        }
+    }
+}
+
+struct SymbolTable {
+    symbols: Vec<HashMap<String, Identifier>>,
+    bitinfo: Vec<HashMap<BitType, IdentifierOrSubscripted>>,
+    reginfo: Vec<HashMap<RegisterType, IdentifierOrSubscripted>>,
+    gates: IndexMap<String, QuantumGateDefinition>,
+    stdgates: HashSet<String>,
+    _counter: Counter,
+}
+
+impl SymbolTable {
+    fn new() -> Self {
+        let symbols = vec![HashMap::new()];
+        let bitinfo = vec![HashMap::new()];
+        let reginfo = vec![HashMap::new()];
+        Self {
+            symbols,
+            bitinfo,
+            reginfo,
+            gates: IndexMap::new(),
+            stdgates: HashSet::new(),
+            _counter: Counter::new(),
+        }
+    }
+
+    fn bind(&mut self, name: &str) -> ExporterResult<()> {
+        if let Some(symbols) = self.symbols.last() {
+            if !symbols.contains_key(name) {
+                self.bind_no_check(name);
+            }
+        } else {
+            return Err(QASM3ExporterError::Error(format!(
+                "Symbol table is empty, cannot bind '{}'",
+                name
+            )));
+        }
+
+        Ok(())
+    }
+
+    fn bind_no_check(&mut self, name: &str) {
+        let id = Identifier {
+            string: name.to_string(),
+        };
+        if let Some(last) = self.symbols.last_mut() {
+            last.insert(name.to_string(), id);
+        }
+    }
```

Comment:
Minor naming: typically if there's two variants of a function, one returns `Result` and one panics instead, it Rust convention to call the `Result` one `try_bind` and the panicking one `bind`.

Actually in this case, though, there should be no need for the `Result` from `bind`: that error represents a non-actionable internal logic error.  If we hit that path, the internal logic of the exporter _must_ be broken, which is a suitable condition for panicking.